chain/client/src/stateless_validation/chunk_validator/mod.rs

```diff
@@ -588,18 +595,44 @@ impl Client {
         witness: ChunkStateWitness,
         peer_id: PeerId,
         processing_done_tracker: Option<ProcessingDoneTracker>,
-    ) -> Result<Option<ChunkStateWitness>, Error> {
+    ) -> Result<(), Error> {
         // TODO(#10502): Handle production of state witness for first chunk after genesis.
         // Properly handle case for chunk right after genesis.
         // Context: We are currently unable to handle production of the state witness for the
         // first chunk after genesis as it's not possible to run the genesis chunk in runtime.
         let prev_block_hash = witness.inner.chunk_header.prev_block_hash();
         let prev_block = match self.chain.get_block(prev_block_hash) {
             Ok(block) => block,
-            Err(_) => {
-                return Ok(Some(witness));
+            Err(Error::DBNotFoundErr(_)) => {
+                // Previous block isn't available at the moment, add this witness to the orphan pool.
+                self.handle_orphan_state_witness(witness)?;
+                return Ok(());
             }
+            Err(err) => return Err(err),
         };
+        self.process_chunk_state_witness_with_prev_block(
+            witness,
+            peer_id,
+            &prev_block,
+            processing_done_tracker,
+        )
+    }
+
+    pub fn process_chunk_state_witness_with_prev_block(
+        &mut self,
+        witness: ChunkStateWitness,
+        peer_id: PeerId,
+        prev_block: &Block,
+        processing_done_tracker: Option<ProcessingDoneTracker>,
+    ) -> Result<(), Error> {
+        if witness.inner.chunk_header.prev_block_hash() != prev_block.hash() {
```

Comment:
this is one example where I think asserting is more appropriate than returning an error
* the contract of this function already assumes block to correspond to `witness.inner.chunk_header.prev_block_hash()`, so violating that is a programmatic error, not an "invalid input" kind of error
* having error handling here might give an impression that this could actually happened under some input, which is not true
* returning error might mask an underlying issue and result in more time spent on debugging or completely hide it which is even worse

Reply:
I don't know, I'm really scared of introducing code that could lead to DoS vulnerability. It might not be a big threat to normal services, but on a blockchain I imagine that a DoS means that it'd be possible to kick out validators and cause all sorts of mayhem.
IMO passing the wrong block to `process_chunk_state_witness` is not a fatal error - it's invalid input, for which the function can safely fail.
I added it because I know that there are plans to process witnesses before the block is applied, and I was worried that someone might accidentally use it incorrectly once the complexity of things increases.

I don't see a reason to risk a panic here. This error will show up in the logs with an `ERROR` log-level, and I think anyone debugging an issue looks for those, so the visibility is there. Maybe we could introduce a `BUG` log-level, that would be used for non-critical bugs?

Reply:
feel free to merge the PR as it is and we can discuss this on protocol core weekly sync